web/modules/custom/dpl_go/src/GoSite.php

```diff
@@ -89,4 +108,50 @@ public function isGoNode(EntityInterface $node): bool {
     return str_starts_with($node->bundle(), 'go_');
   }
 
+  /**
+   * Determine if the given nid is a Go node.
+   */
+  public function isGoNid(string $nid): ?bool {
+    // We store our cache of node types in chunks of 100 to limit the amount of
+    // state items and queries. And we use state instead of cache as a nodes
+    // type cannot be changed after creation anyway, and even CACHE_PERMANENT
+    // entries are cleared on cache rebuild.
+    $cache_num = floor(intval($nid) / 100);
+    $state_id = "dpl_go.node_type_cache_{$cache_num}";
+
+    // Use a static cache. This will drastically limit the amount of queries we
+    // need to do.
+    if (!isset($this->typeCoche[$cache_num])) {
+      $this->typeCoche[$cache_num] = $this->state->get($state_id, []);
+    }
+
+    if (isset($this->typeCoche[$cache_num][$nid])) {
+      return $this->typeCoche[$cache_num][$nid];
+    }
+
+    $this->typeCoche[$cache_num][$nid] = NULL;
+
+    $node = $this->nodeStorage->load($nid);
+
+    if ($node) {
+      $this->typeCoche[$cache_num][$nid] = $this->isGoNode($node);
+    }
+
+    // Add in fresh copy of saved state to guard against race conditions.
+    $this->typeCoche[$cache_num] += $this->state->get($state_id, []);
```

Comment:
Consider if this approach should be reversed e.g.

1. Get the saved state
2. Set the value for the current node in the state

Personally I have always found the behavior of the `+=` operator hard to grok compared to e.g. `array_merge()`.

Reply:
Well, `State` makes it unnecessary.